chemprop/uncertainty/uncertainty_calibrator.py

```diff
@@ -615,6 +621,253 @@ def apply_calibration(self, uncal_predictor: UncertaintyPredictor):
+            )
+
+    @staticmethod
+    def nonconformity_scores(uncal_preds):
+        """Fixed per class. Example is for adaptive.                                                                                                                                                                                                                              
+                                                                                                                                                                                                                                                                              
+        Args:                                                                                                                                                                                                                                                                     
+            preds: [num_examples, num_tasks, num_classes]                                                                                                                                                                                                                         
+                                                                                                                                                                                                                                                                              
+        Returns:                                                                                                                                                                                                                                                                  
+            scores: [num_examples, num_tasks, num_classes]                                                                                                                                                                                                                        
+        """
+        # Sort indices from high to low
+        sort_inds = np.argsort(-uncal_preds, axis=2)
+
+        # Sort and calculate scores
+        sorted_preds = np.take_along_axis(uncal_preds, sort_inds, axis=2)
+        sorted_scores = sorted_preds.cumsum(axis=2)
+
+        # Unsort and return
+        unsort_inds = np.argsort(sort_inds, axis=2)
+        unsorted_scores = np.take_along_axis(sorted_scores, unsort_inds, axis=2)
+        return unsorted_scores
+
+
+class ConformalMultilabelCalibrator(UncertaintyCalibrator):
+    """
+    Conformal Calibrator.
+    """
+    @property
+    def label(self):
+        return f"{self.uncertainty_method}_conformal_multilabel"
+
+    def raise_argument_errors(self):
+        super().raise_argument_errors()
+        if self.dataset_type != "classification":
+            raise ValueError(
+                "Conformal is only implemented for classification dataset types."
+            )
+
+    def calibrate(self):
```

Comment:
Ah, ok, I think you had a slight conceptual mistake here?

When computing
```python
self.tout_hat = np.quantile(calibration_scores_out, self.alpha / 2)
self.tin_hat = np.quantile(calibration_scores_in, 1 - self.alpha / 2)
```
you are trying to directly replace Eq. 15 of alg 3 with a quantile that is independent of x? You can only do this if tout_hat/tin_hat are learned (in this case, merely computed) on separate data (note the use of I_2 vs I_3 in Alg. 3). 

Instead what I was suggesting (to avoid needing to both estimate tout/tin and the associated data for this step), was to just use a much simpler (and, yes, less efficient) estimator:

```python
# Assume that lower nonconformity scores are better. i.e., the lower s_k(x), the more likely y_k = 1.

# Under the null that the new y_k is 0, s_k(x) will be greater than the alpha/2 quantile of min nonconf. scores when y_k is 0, at least 1 - alpha/2 fraction of the time.
# Flip scores to be negative to compute quantile, but this should roughly be the same as
# if s_k(x) > Quantile(alpha/2, min_k s_k(x) : Y_k = 0 \cup {0})
if -s_k(x) ≤ Quantile(1 - alpha/2, -min_k s_k(X) : Y_k = 0 \cup {inf}):
    C_in(k) = 0
else:
    C_in(k) = 1

# Under the null that the new y_k is 1, s_k(x) will be less than the 1-alpha/2 quantile of max nonconf. scores when y_k is 1, at least 1 - alpha/2 fraction of the time.
if s_k(x) ≤ Quantile(1 - alpha/2, max_k s_k(x) : y_k = 1 \cup {inf}):
    C_out(k) = 1
else:
    C_out(k) = 0
```
You can double check me on this. 


To do this efficiently in numpy... follow like this (note use of masking to ignore scores that are not meeting y_k = 0. Similar for y_k = 1.
```
has_zeros = np.sum(targets == 0, axis=1) > 0
inds_zero = targets[has_zeros] == 0  # Dont forget about ignoring the mask Nans too
... scores = compute_scores ... # this will be just -uncal_preds, but might want something else sometime later.
masked_scores = scores * (1 - inds_zero) + np.Inf * inds_zeros
min_zeros = np.min(masked_scores, axis=1)
... compute quantile...
```

Reply:
I see, so essentially just ignoring the qhat step and replacing tin, tout with tin_hat and tout_hat